openedx/core/lib/api/utils.py

```diff
@@ -0,0 +1,41 @@
+import functools
+from importlib import import_module
+
+from django.conf import settings
+
+
+def pluggable_override(override):
+    """
+    This decorator allows overriding any function or method by pointing to an alternative implementation
+    with `override` param.
+    :param override: path to the alternative function
+
+    Usage:
+
+    1. Add this decorator to your original method. `DO_SOMETHING_IMPL` is the variable name in settings that can be
+       used for overriding this method.
+        >>> @pluggable_override('DO_SOMETHING_IMPL')
+        ...     def do_something(self):
+        ...         return _something()
+
+    2. Prepare an alternative implementation:
+        >>> def do_something_different(something):
+        ...     return getattr(something, '_different')()
+
+
+    3. Specify the path in settings (e.g. in `envs/private.py`):
+        >>> DO_SOMETHING_IMPL = 'do_something_different_plugin.do_something.do_something_different'
+    """
+    def decorator(f):
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            function_string = getattr(settings, override)
+            if not function_string:
+                return f(*args, **kwargs)
+
+            module, function = function_string.rsplit('.', 1)
+            mod = import_module(module)
+            func = getattr(mod, function)
+            return func(*args, **kwargs)
```

Comment:
Maybe you should pass `f` into `func` so that `func` can optionally call the base implementation?

Also, it would be good if `DO_SOMETHING_IMPL` can be a list of functions, so that they can be chained. Otherwise, for example, as the platform grows, one might have two plugins installed that both want to add a custom XBlock icon handler. You could do something like:

```
prev_fn = f  # the base method in edx-platform
for impl in override_functions:
    prev_fn = partial(impl, prev_fn)
prev_fn()  # Call the last function in DO_SOMETHING_IMPL, and give it the chance to call the previous one, which can call the previous one, ..., which can call the base implementation. Or any one can return without calling the previous/base implementation.
```

Reply:
That's a great idea!
Could you please check if [this](https://github.com/edx/edx-platform/pull/21433/commits/d67935154f8057d5737b9164de098d3d5074b441#diff-03c7b83e1f4c1cacc306a7966eaaf27cR47-R55) matches your idea, when you have a moment?

Reply:
Looks great, thanks @Agrendalath !